File: Autonomous Mobile Robots Course/Chapter1/Lesson4/Chapter1_Lesson4.py
"""
Chapter1_Lesson4.py
Sensing–Estimation–Navigation pipeline demo (2D unicycle) using weighted least squares (WLS)
prior + range-bearing landmark measurements.
Dependencies: numpy, matplotlib
Optional robotics ecosystem pointers: ROS 2 (rclpy), nav2, GTSAM (not required for this demo).
"""

import numpy as np
import matplotlib.pyplot as plt

def wrap_angle(a: float) -> float:
    """Wrap angle to (-pi, pi]."""
    return (a + np.pi) % (2.0 * np.pi) - np.pi

def unicycle_step(x: np.ndarray, v: float, w: float, dt: float) -> np.ndarray:
    """Discrete-time unicycle integration."""
    x_new = x.copy()
    th = x[2]
    x_new[0] = x[0] + v * dt * np.cos(th)
    x_new[1] = x[1] + v * dt * np.sin(th)
    x_new[2] = wrap_angle(x[2] + w * dt)
    return x_new

def meas_model(x: np.ndarray, landmark: np.ndarray) -> np.ndarray:
    """Range-bearing measurement z = [r, b] to landmark in world coordinates."""
    dx = landmark[0] - x[0]
    dy = landmark[1] - x[1]
    r = np.hypot(dx, dy)
    b = wrap_angle(np.arctan2(dy, dx) - x[2])
    return np.array([r, b])

def meas_jacobian(x: np.ndarray, landmark: np.ndarray) -> np.ndarray:
    """Jacobian H = d h / d x for range-bearing measurement."""
    dx = landmark[0] - x[0]
    dy = landmark[1] - x[1]
    q = dx*dx + dy*dy
    r = np.sqrt(q)

    # Avoid division by zero
    eps = 1e-9
    q = max(q, eps)
    r = max(r, eps)

    # h1 = r = sqrt((lx-x)^2 + (ly-y)^2)
    dr_dx = -(dx) / r
    dr_dy = -(dy) / r
    dr_dth = 0.0

    # h2 = b = atan2(dy, dx) - th
    db_dx =  dy / q
    db_dy = -dx / q
    db_dth = -1.0

    return np.array([[dr_dx, dr_dy, dr_dth],
                     [db_dx, db_dy, db_dth]])

def wls_update_iterated(x_prior: np.ndarray, P_prior: np.ndarray,
                        z_list: list, landmarks: np.ndarray,
                        R: np.ndarray, iters: int = 2) -> tuple[np.ndarray, np.ndarray]:
    """
    Iterated WLS (Gauss-Newton) for a prior + stacked measurements.
    Minimizes:
        (x-x_prior)^T P_prior^{-1} (x-x_prior) + sum_i (z_i - h_i(x))^T R^{-1} (z_i - h_i(x))
    """
    if len(z_list) == 0:
        return x_prior, P_prior

    # Stack measurements
    z = np.concatenate(z_list, axis=0)  # shape (2m,)
    m = len(z_list)

    # Block-diagonal measurement covariance
    R_big = np.kron(np.eye(m), R)  # (2m x 2m)
    Rinv = np.linalg.inv(R_big)

    Pinv = np.linalg.inv(P_prior)

    x = x_prior.copy()
    for _ in range(iters):
        # Build stacked residual and Jacobian
        h = []
        H = []
        for i in range(m):
            hi = meas_model(x, landmarks[i])
            h.append(hi)
            H.append(meas_jacobian(x, landmarks[i]))
        h = np.concatenate(h, axis=0)
        H = np.vstack(H)

        # Residual (bearing residual must be wrapped)
        res = z - h
        for i in range(m):
            res[2*i + 1] = wrap_angle(res[2*i + 1])

        # Solve normal equations for delta:
        # (P^{-1} + H^T R^{-1} H) delta = H^T R^{-1} res + P^{-1}(x_prior - x)
        A = Pinv + H.T @ Rinv @ H
        b = H.T @ Rinv @ res + Pinv @ (x_prior - x)
        delta = np.linalg.solve(A, b)

        x = x + delta
        x[2] = wrap_angle(x[2])

    # Posterior covariance at final linearization point
    h = []
    H = []
    for i in range(m):
        h.append(meas_model(x, landmarks[i]))
        H.append(meas_jacobian(x, landmarks[i]))
    H = np.vstack(H)
    P_post = np.linalg.inv(Pinv + H.T @ Rinv @ H)
    return x, P_post

def main():
    np.random.seed(7)

    # Map (known landmarks): choose 3 well-spread landmarks
    landmarks = np.array([[5.0, 0.0],
                          [6.0, 6.0],
                          [0.0, 6.0]])

    dt = 0.1
    N = 350

    # True state and estimate (x, y, theta)
    x_true = np.array([0.0, 0.0, 0.0])
    x_hat  = np.array([0.0, 0.0, 0.0])

    # Prior covariance
    P = np.diag([0.15**2, 0.15**2, (np.deg2rad(8.0))**2])

    # Noise levels
    sigma_v = 0.08          # m/s odometry noise
    sigma_w = np.deg2rad(3) # rad/s odometry noise
    sigma_r = 0.12          # m range noise
    sigma_b = np.deg2rad(2) # rad bearing noise

    # Navigation goal and controller gains (simple go-to-goal)
    goal = np.array([7.0, 7.0])
    k_heading = 1.6
    v_max = 0.9

    # Measurement covariance per landmark measurement
    R = np.diag([sigma_r**2, sigma_b**2])

    hist_true = np.zeros((N, 3))
    hist_hat  = np.zeros((N, 3))

    for k in range(N):
        # --- Navigation (policy uses estimate) ---
        dx = goal[0] - x_hat[0]
        dy = goal[1] - x_hat[1]
        dist = np.hypot(dx, dy)
        desired_heading = np.arctan2(dy, dx)
        heading_error = wrap_angle(desired_heading - x_hat[2])

        v_cmd = v_max * np.tanh(dist)      # smoothly saturating speed
        w_cmd = k_heading * heading_error  # proportional heading control

        # --- True motion (includes unmodeled disturbances) ---
        # Model a mild lateral slip disturbance as small random heading perturbation.
        slip = np.random.normal(0.0, np.deg2rad(0.35))
        x_true = unicycle_step(x_true, v_cmd, w_cmd + slip/dt, dt)

        # --- Sensing: odometry (control-derived) ---
        v_meas = v_cmd + np.random.normal(0.0, sigma_v)
        w_meas = w_cmd + np.random.normal(0.0, sigma_w)

        # --- Estimation: prediction from odometry ---
        x_prior = unicycle_step(x_hat, v_meas, w_meas, dt)

        th = x_hat[2]
        F = np.array([[1.0, 0.0, -v_meas*dt*np.sin(th)],
                      [0.0, 1.0,  v_meas*dt*np.cos(th)],
                      [0.0, 0.0,  1.0]])

        # Simple process noise model (odometry-induced)
        Q = np.diag([(sigma_v*dt)**2, (sigma_v*dt)**2, (sigma_w*dt)**2])
        P_prior = F @ P @ F.T + Q

        # --- Sensing: landmark measurements (range-bearing) ---
        z_list = []
        # Use all landmarks within range to emulate a perception module
        max_range = 8.0
        for lm in landmarks:
            z_true = meas_model(x_true, lm)
            if z_true[0] <= max_range:
                z_noisy = z_true + np.array([np.random.normal(0.0, sigma_r),
                                             np.random.normal(0.0, sigma_b)])
                z_noisy[1] = wrap_angle(z_noisy[1])
                z_list.append(z_noisy)

        # --- Estimation: iterated WLS correction (prior + measurements) ---
        x_hat, P = wls_update_iterated(x_prior, P_prior, z_list, landmarks[:len(z_list)], R, iters=2)

        hist_true[k] = x_true
        hist_hat[k]  = x_hat

    # --- Visualization ---
    plt.figure()
    plt.plot(hist_true[:,0], hist_true[:,1], label="true")
    plt.plot(hist_hat[:,0],  hist_hat[:,1],  label="estimated")
    plt.scatter(landmarks[:,0], landmarks[:,1], marker="x", label="landmarks")
    plt.scatter(goal[0], goal[1], marker="*", label="goal")
    plt.axis("equal")
    plt.grid(True)
    plt.legend()
    plt.title("Sensing–Estimation–Navigation loop (demo)")
    plt.show()

if __name__ == "__main__":
    main()

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter1/Lesson4/Chapter1_Lesson4.m
% Chapter1_Lesson4.m
% Sensing–Estimation–Navigation pipeline demo (2D unicycle) using WLS (prior + range-bearing).
% Requires: base MATLAB. Optional (not required): Robotics System Toolbox / Simulink for later chapters.
% Run:
%   Chapter1_Lesson4

clear; clc; close all;
rng(7);

wrapAngle = @(a) mod(a + pi, 2*pi) - pi;

unicycleStep = @(x,v,w,dt) [ ...
    x(1) + v*dt*cos(x(3)); ...
    x(2) + v*dt*sin(x(3)); ...
    wrapAngle(x(3) + w*dt) ];

measModel = @(x,lm) [ ...
    hypot(lm(1)-x(1), lm(2)-x(2)); ...
    wrapAngle(atan2(lm(2)-x(2), lm(1)-x(1)) - x(3)) ];

measJacobian = @(x,lm) localMeasJacobian(x,lm);

% Map (known landmarks)
landmarks = [5 0; 6 6; 0 6];

dt = 0.1; N = 350;

xTrue = [0;0;0];
xHat  = [0;0;0];

P = diag([0.15^2, 0.15^2, deg2rad(8)^2]);

sigmaV = 0.08;
sigmaW = deg2rad(3);
sigmaR = 0.12;
sigmaB = deg2rad(2);

goal = [7;7];
kHeading = 1.6;
vMax = 0.9;

R = diag([sigmaR^2, sigmaB^2]);

histTrue = zeros(N,3);
histHat  = zeros(N,3);

for k=1:N
    % --- Navigation (policy uses estimate) ---
    dx = goal(1) - xHat(1);
    dy = goal(2) - xHat(2);
    dist = hypot(dx,dy);
    desired = atan2(dy,dx);
    headingErr = wrapAngle(desired - xHat(3));
    vCmd = vMax*tanh(dist);
    wCmd = kHeading*headingErr;

    % --- True motion with mild slip disturbance ---
    slip = deg2rad(0.35)*randn;
    xTrue = unicycleStep(xTrue, vCmd, wCmd + slip/dt, dt);

    % --- Odometry ---
    vMeas = vCmd + sigmaV*randn;
    wMeas = wCmd + sigmaW*randn;

    % --- Prediction ---
    xPrior = unicycleStep(xHat, vMeas, wMeas, dt);

    th = xHat(3);
    F = [1 0 -vMeas*dt*sin(th);
         0 1  vMeas*dt*cos(th);
         0 0  1];

    Q = diag([(sigmaV*dt)^2, (sigmaV*dt)^2, (sigmaW*dt)^2]);
    PPrior = F*P*F' + Q;

    % --- Landmark measurements ---
    zList = [];
    lmList = [];
    maxRange = 8.0;
    for j=1:size(landmarks,1)
        zTrue = measModel(xTrue, landmarks(j,:)');
        if zTrue(1) <= maxRange
            zNoisy = zTrue + [sigmaR*randn; sigmaB*randn];
            zNoisy(2) = wrapAngle(zNoisy(2));
            zList = [zList; zNoisy];
            lmList = [lmList; landmarks(j,:)];
        end
    end

    % --- Iterated WLS correction (2 iterations) ---
    [xHat, P] = wlsUpdateIterated(xPrior, PPrior, zList, lmList, R, 2, wrapAngle, measModel, measJacobian);

    histTrue(k,:) = xTrue';
    histHat(k,:)  = xHat';
end

figure;
plot(histTrue(:,1), histTrue(:,2), 'LineWidth', 1.5); hold on;
plot(histHat(:,1), histHat(:,2),  'LineWidth', 1.5);
scatter(landmarks(:,1), landmarks(:,2), 70, 'x', 'LineWidth', 2);
scatter(goal(1), goal(2), 90, '*', 'LineWidth', 2);
axis equal; grid on;
legend('true','estimated','landmarks','goal');
title('Sensing–Estimation–Navigation loop (demo)');

% ===== Local functions =====

function H = localMeasJacobian(x,lm)
    dx = lm(1) - x(1);
    dy = lm(2) - x(2);
    q = dx^2 + dy^2;
    q = max(q, 1e-12);
    r = sqrt(q);
    r = max(r, 1e-12);

    H = zeros(2,3);
    H(1,1) = -dx/r;
    H(1,2) = -dy/r;
    H(1,3) = 0;

    H(2,1) =  dy/q;
    H(2,2) = -dx/q;
    H(2,3) = -1;
end

function [xPost, PPost] = wlsUpdateIterated(xPrior, PPrior, zList, lmList, R, iters, wrapAngle, measModel, measJacobian)
    if isempty(zList)
        xPost = xPrior;
        PPost = PPrior;
        return;
    end

    m = size(zList,1);
    z = reshape(zList', [], 1); % (2m x 1)

    Rbig = kron(eye(m), R);
    Rinv = inv(Rbig);
    Pinv = inv(PPrior);

    x = xPrior;
    for it = 1:iters
        h = zeros(2*m,1);
        H = zeros(2*m,3);
        for j=1:m
            lm = lmList(j,:)';
            hj = measModel(x,lm);
            h(2*j-1:2*j) = hj;
            H(2*j-1:2*j,:) = measJacobian(x,lm);
        end

        res = z - h;
        for j=1:m
            res(2*j) = wrapAngle(res(2*j));
        end

        A = Pinv + H'*Rinv*H;
        b = H'*Rinv*res + Pinv*(xPrior - x);

        delta = A \ b;
        x = x + delta;
        x(3) = wrapAngle(x(3));
    end

    % Posterior covariance at final linearization point
    H = zeros(2*m,3);
    for j=1:m
        lm = lmList(j,:)';
        H(2*j-1:2*j,:) = measJacobian(x,lm);
    end
    PPost = inv(Pinv + H'*Rinv*H);
    xPost = x;
end
